### PR TITLE
added 2s-AGCN implementation

### PR DESCRIPTION
**Pull request by alexriedel1**
Added this implementation https://github.com/lshiwjx/2s-AGCN of Two-Stream Adaptive Graph Convolutional Networks for Skeleton-Based Action Recognition https://arxiv.org/abs/1805.07694

//fixing spellings and adding test...

### CODE REVIEW

**Comment by alexriedel1** [comment]
Don't be too harsh it's my first pull request on github ever 😂

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
+import math
+
+import numpy as np
+import torch
+import torch.nn as nn
+from torch.autograd import Variable
+
+
+def edge2mat(link, num_node):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:9]: Can you write this in pure PyTorch?

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    A = np.zeros((num_node, num_node))
+    for i, j in link:
+        A[j, i] = 1
+    return A
+
+
+def normalize_digraph(A):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:16]: Can you rewrite this in PyTorch?

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+    return A
+
+def edge_index2inward(edge_index):
+    inward = []
+    for first_row, second_row in zip(edge_index[0], edge_index[1]):
+        inward.append((first_row, second_row))
+    return inward
+
+class Graph:
+    r"""
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:40]: Can you rewrite this with pure PyTorch?

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    return AD
+
+def get_spatial_graph(num_node, self_link, inward, outward):
+    I = edge2mat(self_link, num_node)
+    In = normalize_digraph(edge2mat(inward, num_node))
+    Out = normalize_digraph(edge2mat(outward, num_node))
+    A = np.stack((I, In, Out))
+    return A
+
+def edge_index2inward(edge_index):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:33]: Can you rewrite this with PyTorch functions?

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        self.relu = nn.ReLU(inplace=True)
+        conv_init(self.conv)
+        bn_init(self.bn, 1)
+
+    def forward(self, x):
+        x = self.bn(self.conv(x))
+        return x
+
+
+class unit_gcn(nn.Module):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:106]: Class names have camel cases.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    nn.init.kaiming_normal_(conv.weight, mode='fan_out')
+    nn.init.constant_(conv.bias, 0)
+
+
+def bn_init(bn, scale):
+    nn.init.constant_(bn.weight, scale)
+    nn.init.constant_(bn.bias, 0)
+
+
+class unit_tcn(nn.Module):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:80]: Class names have camelcasing - UnitTCN

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+def conv_branch_init(conv, branches):
+    weight = conv.weight
+    n = weight.size(0)
+    k1 = weight.size(1)
+    k2 = weight.size(2)
+    nn.init.normal_(weight, 0, math.sqrt(2. / (n * k1 * k2 * branches)))
+    nn.init.constant_(conv.bias, 0)
+
+
+def conv_init(conv):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:70]: Move init to private class method

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    nn.init.kaiming_normal_(conv.weight, mode='fan_out')
+    nn.init.constant_(conv.bias, 0)
+
+
+def bn_init(bn, scale):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:75]: Move init to private class method.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        self.neighbor = self.inward + self.outward
+
+        self.A = self.get_adjacency_matrix()
+
+    def get_adjacency_matrix(self):
+        A = get_spatial_graph(self.num_nodes, self.self_link, self.inward, self.outward)
+
+        return A
+
+def conv_branch_init(conv, branches):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:61]: Move init to private class method.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    def __init__(self, edge_index:list, num_nodes:int):
+        self.num_nodes = num_nodes
+        self.inward = edge_index2inward(edge_index)
+        self.self_link = [(i, i) for i in range(num_nodes)]
+        self.outward = [(j, i) for (i, j) in self.inward]
+        self.neighbor = self.inward + self.outward
+
+        self.A = self.get_adjacency_matrix()
+
+    def get_adjacency_matrix(self):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:56]: Please specify type of return value.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    Dl = np.sum(A, 0)
+    h, w = A.shape
+    Dn = np.zeros((w, w))
+    for i in range(w):
+        if Dl[i] > 0:
+            Dn[i, i] = Dl[i] ** (-1)
+    AD = np.dot(A, Dn)
+    return AD
+
+def get_spatial_graph(num_node, self_link, inward, outward):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:26]: Please rewrite in PyTorch.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+        out_channels (int): Number of output features.
+        A (Tensor array): Adaptive Graph.
+        coff_embedding (int, optional): Coefficient Embeddings
+        num_subset (int, optional): Subsets for adaptive graphs, see :math:`\mathbf{A}, \mathbf{B}, \mathbf{C}` in https://arxiv.org/abs/1805.07694 for details
+        adaptive (bool, optional): Apply Adaptive Graph Convolutions.
+        attention (bool, optional): Apply Attention.
+    """
+    def __init__(self, in_channels:int, out_channels:int, A:torch.FloatTensor, coff_embedding:int=4, num_subset:int=3, adaptive: bool=True, attention: bool=True):
+        super(unit_gcn, self).__init__()
+        inter_channels = out_channels // coff_embedding
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:122]: Please break up the init to submethods. Use private submethods for meaningful subunits of code.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+    Args:
+        in_channels (int): Number of input features.
+        out_channels (int): Number of output features.
+        A (Tensor array): Adaptive Graph.
+        coff_embedding (int, optional): Coefficient Embeddings
+        num_subset (int, optional): Subsets for adaptive graphs, see :math:`\mathbf{A}, \mathbf{B}, \mathbf{C}` in https://arxiv.org/abs/1805.07694 for details
+        adaptive (bool, optional): Apply Adaptive Graph Convolutions.
+        attention (bool, optional): Apply Attention.
+    """
+    def __init__(self, in_channels:int, out_channels:int, A:torch.FloatTensor, coff_embedding:int=4, num_subset:int=3, adaptive: bool=True, attention: bool=True):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:120]: Please break up the line with linting - https://pypi.org/project/pylint/.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                conv_init(m)
+            elif isinstance(m, nn.BatchNorm2d):
+                bn_init(m, 1)
+        bn_init(self.bn, 1e-6)
+        for i in range(self.num_subset):
+            conv_branch_init(self.conv_d[i], self.num_subset)
+
+    def forward(self, x):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:193]: Please add private methods to break up the forward pass.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+            se = y.mean(-1).mean(-1)
+            se1 = self.relu(self.fc1c(se))
+            se2 = self.sigmoid(self.fc2c(se1))
+            y = y * se2.unsqueeze(-1).unsqueeze(-1) + y
+
+        return y
+
+
+
+class TCN_GCN_unit(nn.Module):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:241]: Add Camelcasing name.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+        in_channels (int): Number of input features.
+        out_channels (int): Number of output features.
+        edge_index (PyTorch LongTensor): Graph edge indices.
+        num_nodes (int): Number of nodes in the network.
+        stride (int, optional): Time strides during temporal convolution.
+        residual (bool, optional): Applying connections.
+        adaptive (bool, optional): Applying adaptive graphs
+        attention (bool, optional): Applying Attention
+    """
+    def __init__(self, in_channels:int, out_channels:int, edge_index:torch.LongTensor, num_nodes:int, stride:int=1, residual:bool=True, adaptive:bool=True, attention:bool=True):
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:258]: Please break the line up. Hard to read.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+        super(TCN_GCN_unit, self).__init__()
+        self.edge_index = edge_index
+        self.num_nodes = num_nodes
+
+        self.graph = Graph(self.edge_index, self.num_nodes)
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:263]: Remove multiple spaces.

File: torch_geometric_temporal/nn/attention/2sagcn.py [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+        Return types:
+            * **X** (PyTorch FloatTensor)* - Sequence of node features, with shape (B, out_channels, T_in//stride, N_nodes).
+        """
+
+        y = self.relu(self.tcn1(self.gcn1(x)) + self.residual(x))
+
+        return y
+
+
+if __name__ == "__main__":
```
Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:297]: We write separate test for doing some sanity checks in the test folder. Test method names start with test. This way we can check the coverage of the test.

Comment (benedekrozemberczki) [inline torch_geometric_temporal/nn/attention/2sagcn.py:297]: Please remove this.

**Review by benedekrozemberczki — CHANGES_REQUESTED** [review]
Dear Alex, 

🎇 Thank you for contributing! It is really appreciated, do not worry - it is a really high quality contribution 👍. Please check the following things (you will have to add modifications in a number of places, not just the py file):

- Modify the init file of attention models.
- Run through your py file pylint.
- The docstring should contain default hyperparameter values e.g. kernel_size: int=9.
- Please write tests in the test folder.
- Add the paper to the readme - give reference to the autodocs.
- Add a reference to the paper in docs/source/notes/resources.rst.
- Add a reference to the layer in docs/source/modules/root.rst.

We are going to submit the design paper to the CIKM resource track. If you would like to contribute other models (additional 1-2) we are happy to add you as a co-author.

Bests,

Benedek

**Comment by alexriedel1** [comment]
> Dear Alex,
> 
> 🎇 Thank you for contributing! It is really appreciated, do not worry - it is a really high quality contribution 👍. Please check the following things (you will have to add modifications in a number of places, not just the py file):
> 
> * Modify the init file of attention models.
> * Run through your py file pylint.
> * The docstring should contain default hyperparameter values e.g. kernel_size: int=9.
> * Please write tests in the test folder.
> * Add the paper to the readme - give reference to the autodocs.
> * Add a reference to the paper in docs/source/notes/resources.rst.
> * Add a reference to the layer in docs/source/modules/root.rst.
> 
> We are going to submit the design paper to the CIKM resource track. If you would like to contribute other models (additional 1-2) we are happy to add you as a co-author.
> 
> Bests,
> 
> Benedek

Alright thanks a lot! I've just commited the required changes. I'd be happy to add some more models soon and being considered as co-author :)

**Comment by benedekrozemberczki** [comment]
Dear Alex,

Merged the PR! There was a minor fix needed for the test, but now it runs with CI and the docs also pass. We will have a research meeting on Tuesday, about the paper submission. If you drop me an e-mail at: benedek.rozemberczki@gmail.com I can send you an invite. Do you have other models in mind?

Bests,

Benedek

**Comment by alexriedel1** [comment]
> Dear Alex,
> 
> Merged the PR! There was a minor fix needed for the test, but now it runs with CI and the docs also pass. We will have a research meeting on Tuesday, about the paper submission. If you drop me an e-mail at: [benedek.rozemberczki@gmail.com](mailto:benedek.rozemberczki@gmail.com) I can send you an invite. Do you have other models in mind?
> 
> Bests,
> 
> Benedek

Thanks Benedek! I'll send you an email, plus I'm planning to add this skeleton based action recognition model MS-3GD https://arxiv.org/abs/2003.14111 (one of the highest scoring in benchmarks for graph based action recognition that also make sense to apply in other tasks...)

**Pull request merged**